File: Maths/kadane_algo.py
# Kadane's algo - O(N) complexity

class Solution(object):
    def kadane(self,arr:list)->int:
        self.best = 0
        self.sum = 0
        self.n = len(arr)
        for i in range(self.n):
            self.sum = max(arr[i],self.sum+arr[i])
            self.best = max(self.best,self.sum)
        return self.best
    

if __name__=='__main__':
    test = [-1,2,4,-3,5,2,-5,2]
    s = Solution()
    print(s.kadane(test)) # 10 [2,4,-3,5,2]

File: Maths/kadane_algo.cpp
/*Kadane's Algorithm - Max sum sub array*/

#include <bits/stdc++.h>
using namespace std;

class Answer{
public:
  static int max_sum(vector<int> array){
    int best = 0, sum = 0;
    for (auto i = 0; array[i]; i++){
     sum = max(array[i],sum+array[i]);
     best = max(best,sum);
    }
    return best;
  }
};

int main(){
  vector<int> v = {-1,2,4,-3,5,2,-5,2};
  cout << Answer::max_sum(v) << "\n";
}
